Mathlib/RingTheory/Coalgebra/GroupLike.lean

```diff
@@ -0,0 +1,155 @@
+Authors: Yaël Dillies, Michał Mrugała
+-/
+import Mathlib.RingTheory.Coalgebra.Equiv
+import Mathlib.RingTheory.Flat.Domain
+
+/-!
+# Group-like elements in a coalgebra
+
+This file defines group-like elements in a coalgebra, i.e. elements `a` such that `ε a = 1` and
+`Δ a = a ⊗ₜ a`.
+
+## Main declarations
+
+* `IsGroupLikeElem`: Predicate for an element in a coalgebra to be group-like.
+* `linearIndepOn_isGroupLikeElem`: Group-like elements over a domain are linearly independent.
+-/
+
+open Coalgebra Function TensorProduct
+
+variable {F R A B : Type*}
+
+section CommSemiring
+variable [CommSemiring R] [AddCommMonoid A] [AddCommMonoid B] [Module R A] [Coalgebra R A]
+  [Module R B] [Coalgebra R B] {a b : A}
+
+variable (R) in
+/-- A group-like element in a coalgebra is an element `a` such that $\varepsilon(a) = 1$ and
+$\Delta(a) = a ⊗ a$. -/
+@[mk_iff]
+structure IsGroupLikeElem (a : A) where
+  counit_eq_one : counit (R := R) a = 1
+  comul_eq_tmul_self : comul a = a ⊗ₜ[R] a
+
+attribute [simp] IsGroupLikeElem.counit_eq_one IsGroupLikeElem.comul_eq_tmul_self
+
+@[simp] lemma isGroupLikeElem_self {r : R} : IsGroupLikeElem R r ↔ r = 1 := by
+  simp +contextual [isGroupLikeElem_iff]
+
+lemma IsGroupLikeElem.ne_zero [Nontrivial R] (ha : IsGroupLikeElem R a) : a ≠ 0 := by
+  rintro rfl; simpa using ha.counit_eq_one
+
+/-- A coalgebra homomorphism sends group-like elements to group-like elements. -/
+lemma IsGroupLikeElem.map [FunLike F A B] [CoalgHomClass F R A B] (f : F)
+    (ha : IsGroupLikeElem R a) : IsGroupLikeElem R (f a) where
+  counit_eq_one := by rw [CoalgHomClass.counit_comp_apply, ha.counit_eq_one]
+  comul_eq_tmul_self := by rw [← CoalgHomClass.map_comp_comul_apply, ha.comul_eq_tmul_self]; simp
+
+/-- A coalgebra isomorphism preserves group-like elements. -/
+@[simp] lemma isGroupLikeElem_map [EquivLike F A B] [CoalgEquivClass F R A B] (f : F) :
+    IsGroupLikeElem R (f a) ↔ IsGroupLikeElem R a where
+  mp ha := by
+    rw [← (CoalgEquivClass.toCoalgEquiv f).symm_apply_apply a]
+    exact ha.map (CoalgEquivClass.toCoalgEquiv f).symm
+  mpr := .map f
+
+variable (R A) in
+/-- The type of group-like elements in a coalgebra. -/
+@[ext]
+structure GroupLike where
+  /-- The underlying element of a group-like element. -/
+  val : A
+  isGroupLikeElem_val : IsGroupLikeElem R val
+
+namespace GroupLike
+
+attribute [coe] val
+
+instance instCoeOut : CoeOut (GroupLike R A) A where coe := val
+
+attribute [simp] isGroupLikeElem_val
+
+lemma val_injective : Injective (val : GroupLike R A → A) := by rintro ⟨a, ha⟩; congr!
+
+@[simp, norm_cast] lemma val_inj {a b : GroupLike R A} : a.val = b.val ↔ a = b :=
+  val_injective.eq_iff
+
+/-- Identity equivalence between `GroupLike R A` and `{a : A | IsGroupLikeElem R a}`. -/
+def valEquiv : GroupLike R A ≃ Subtype (IsGroupLikeElem R : A → Prop) where
+  toFun a := ⟨a.1, a.2⟩
+  invFun a := ⟨a.1, a.2⟩
+  left_inv _ := rfl
+  right_inv _ := rfl
```

Comment:
```suggestion
@[coe] def val : GroupLike R A → A := Subtype.val

instance instCoeOut : CoeOut (GroupLike R A) A where coe := val

@[simp] theorem isGroupLikeElem_val (x : GroupLike R A) : IsGroupLikeElem R (x : A) := x.2

lemma val_injective : Injective (val : GroupLike R A → A) := Subtype.val_injective

@[ext] theorem ext {x y : GroupLike R A} (h : (x : A) = (y : A)) : x = y := Subtype.ext h

@[simp, norm_cast] lemma val_inj {a b : GroupLike R A} : a.val = b.val ↔ a = b :=
  Subtype.val_inj

/-- Identity equivalence between `GroupLike R A` and `{a : A | IsGroupLikeElem R a}`. -/
def valEquiv : GroupLike R A ≃ Subtype (IsGroupLikeElem R : A → Prop) := Equiv.refl _
```

I hope you would consider this to be an advantage of using Subtype as definition.

Reply:
Sure, then? That happens in Toric!